book/html.md

```diff
@@ -5,408 +5,559 @@ prev: text
-    end with a semicolon \"`;`\"; they resolve to particular characters.
-    Implement support for `&amp;`, `&lt;`, `&gt;`, and `&quot;`, which
-    expand to ampersand, less than, greater than, and the quotation
-    mark. Should you handle entities in the lexer, the parser, or in an
-    earlier pass? Consider that attribute values can contain entities.
--   For some tags, it doesn\'t make sense to have one inside the other.
-    For example, it\'s not clear what it would mean for one paragraph to
-    contain another, so the most common reason for this to happen in a
-    web page is that someone forgot a close tag. Change the parser so
-    that a document like `<p>hello<p>world</p>` results in two sibling
-    nodes instead of one paragraph inside another.
--   The attribute parser doesn\'t correctly handle attribute values
-    that contain spaces, which is valid when the attribute is quoted.
-    Fix this case in the attribute parser. You will likely need to
-    loop over the attribute character-by-character.
-
-[^1]: To be clear: there is additional structure, we\'re just ignoring
-    it for now.
-
-[^2]: Tracking the parent like I\'m doing here is going to be useful for
-    parsing.
-
-[^3]: Yes, it\'s a crazy system, and for a few years in the early '00s
-    the W3C tried to [do away with it](https://www.w3.org/TR/xhtml1/).
-    They failed.
-
-[^4]: Some people put a slash at the end of a self-closing tag (like
-    `<br/>`) but they don\'t have to: `<br>` is self-closing with and
-    without that slash.
-
-[^5]: Where `href` stands for \"hypertext reference\".
-
-[^6]: If you implemented earlier exercises like support for `<a>`,
-    `<small>`, and `<big>`, you will likely have more parts to the
-    state.
+*Entities:* Implement support for `&amp;`, `&lt;`, `&gt;`, and
+`&quot;`, which expand to ampersand, less than, greater than, and the
+quotation mark in HTML. These "entities" can go either in ordinary
+text or inside tags. Should you handle entities in the lexer or the
+parser?
+
+*Comments:* Update the HTML lexer to support comments. Comments in
+HTML begin with `<!--` and end with `-->`. However, comments aren't
+the same as tags: they can contain any text, including left and right
+angle brackets. The lexer should skip comments, not generating any
+token at all. Test: is `<!-->` a comment, or does it just start one?
+
+*Paragraphs:* Since it's not clear what it would mean for one
+paragraph to contain another, the most common reason for this to
+happen in a web page is that someone forgot a close tag. Change the
+parser so that a document like `<p>hello<p>world</p>` results in two
+sibling paragraphs instead of one paragraph inside another.
+
+*Scripts:* JavaScript code embedded in a `<script>` tag uses the left
+angle bracket to mean less-than. Modify your lexer so that the
+contents of `<script>` tags are treated specially: no tags are allowed
+inside `<script>`, except the `</script>` close tag.
```

Comment:
Fun fact: there are some weird restrictions on presence of the string "</script>" in an inline script, for reasons of preventing the HTML parser from needing to understand script syntaxes.

Reply:
Technically it's just `</script` followed by a space, tab, \v, \r, slash, or greater than sign. Yeah, I've run into this once or twice and it's a fun edge case. I talk about it in a video.

https://html.spec.whatwg.org/multipage/parsing.html#script-data-end-tag-name-state